### PR TITLE
feat(testbucket): Phase B — wire go test -json capture into the unit-test lanes, add the master-only record job

### PR DESCRIPTION
**Pull request by invakid404**
**Phase B of the test-bucketing graph (#656).** Instruments the **existing** unit-test jobs to emit timings, adds the write side of the rolling store, and reports the whale divisibility measurement that feeds the go/no-go gate before Phase C. No job is replaced, no test moves lanes — only timing capture is new.

## Divisibility measurement — the gate's answer, and it inverts the design's assumption

Both whales measured at `-count=1` and `-count=10` with per-runnable resolution, fitted to `elapsed(c) = fixed + c × per_iteration`:

| package | `-count=1` | `-count=10` | ratio | projected `-count=100` |
|---|---:|---:|---:|---:|
| `bamlutils/llmhttp` | 10.72s | 83.75s | 7.8× | **13.6 min** |
| `internal/debaml` | 81.76s | 149.05s | 1.8× | **13.7 min** |

Each is dominated by **one runnable that scales per iteration**:

| package | dominant runnable | share | @`-count=100` | why |
|---|---|---:|---:|---|
| `llmhttp` | `TestExactStreamNoGoroutineLeak` | **50%** | 407s | real 2s network waits per iteration |
| `debaml` | `TestConstraintStateCollectorIsTestOnly` | **44%** | 361s | source guard, re-parses the repo every iteration |

**`-run` name-slicing cannot harpoon either.** Its makespan floors at that single test — **6.8 min** for llmhttp, **6.0 min** for debaml — no matter how many slices are cut.

**Count-sharding divides both cleanly**, because it divides *iterations* rather than the test list, and both whales are overwhelmingly per-iteration at `-count=100`:

| S | llmhttp/shard | agg. CPU | debaml/shard | agg. CPU |
|---:|---:|---:|---:|---:|
| 4 | 3.4 min | +1.0% | 4.4 min | +27% |
| **6** | **2.3 min** | **+3.6%** | **3.4 min** | **+47%** |
| 8 | 1.8 min | +6.2% | 2.9 min | +67% |

debaml pays more because it carries ~74s of **once-per-binary** work (`sync.Once` witnesses, chiefly a depth-1000 JSON nesting probe) that count-sharding necessarily repeats in every shard.

**Verdict: CLEANLY DIVISIBLE — divisibility gate GO (owner, 2026-08-25)** — but by the mechanism the design treated as the *zero-data fallback* rather than the *principled* one.

> The **qualitative** conclusion is what the gate turns on and it is settled: each whale's dominant runnable repeats under `-count`, so no name-partition can escape it. Every **absolute** figure here is a local Apple-M4 projection, **pending the first master `-race -count=100` Ubuntu capture** — the tool measures real CI timings via the `record` job and re-balances on demand, so that data, not this PR, drives the split. Neither whale is dominated by an indivisible monolithic test; both are dominated by a **repeatable** one, which is exactly the case count-sharding handles and `-run` slicing does not. Full numbers, method and caveats in `/tmp/shared/tb-b-impl.md`.

## What's wired

**Capture** (`scripts/testbucket-capture.sh`) — both heavy lanes (`unit-tests`' serial module loop and the `bamlutils-race` lane) run each `go test` through a wrapper that adds `-json`, tees the stream to `$RUNNER_TEMP`, and **replays the stream's output events so the console log still reads exactly like a plain `go test` log**. Not decoration: `go test` has no dual-output mode, so a naive `-json` would replace the test log with NDJSON and make every future CI failure harder to read. Events go to `$RUNNER_TEMP`, never the repo, because the two fresh-tree gates later in that job fail on a dirty tree.

The wrapper's sharpest hazard is **fail-open** — a pipeline's exit status is its *last* command's, so without an explicit `PIPESTATUS` read a red suite would be reported through the renderer's happy exit and the job would go **green on failing tests**. Its test suite opens with that case, plus a build failure, and runs in CI on every PR.

**Record** (`record-timings`, master-only per decision 3) — ingests the uploaded events and persists `test-timings.json` to the Actions cache. Per-commit key so each master run writes a fresh immutable entry; prefix `restore-keys` hands the newest previous entry to whoever asks, which is what rolls the weights forward through EWMA instead of restarting. Runs under `!cancelled()` rather than `always()`: a red run still contributes what it measured (ingest drops failed packages, keeps the ones that passed), while a cancelled run's arbitrary partial captures are skipped.

**Restore** (`timings-restore`, every run incl. PRs) — proves decision 3's two rot mitigations continuously rather than once at review time: a PR restores what master recorded, and a **total cache miss still produces a valid, coverage-complete matrix** instead of a failed job. The never-drop-a-test gate inside `plan` is what makes a green line there a statement about coverage rather than exit codes.

## Proof

Rehearsed end to end locally before pushing: the real module loop produced seven event files plus the bamlutils lane, **30,046 events** ingested into a store covering **54/54 live packages**, and `plan` then reported 54 loaded / 0 missing with a complete matrix.

`gofmt` · `shellcheck` (clean) · `CGO_ENABLED=0 go build ./...` · `go vet ./...` · `go test ./cmd/testbucket/...` (95 tests, also under `-race`) · capture-script suite · workflow YAML parses.

## Two findings worth your attention

1. **`internal/debaml` now measures the same size as `llmhttp`** (~13.7 min projected), roughly **double** the design's ~5–8 min estimate. The unit-tests job's root-module invocation carries it under a 20-minute per-invocation timeout.
2. **debaml's heavy tests are source guards that re-parse the tree 100×** for no flake-detection value. Exempting them from the `-count=100` sweep would beat any amount of bucketing.

## Review fix rounds — `d07989ce`, `bc013be8`

**Round 1** replaced the `≥50% named time` upgrade heuristic with a direct comparison of the two mechanisms (`-run` is chosen only when the heaviest name fits under a count-shard's cost), and made the split width **K**. This mattered because the tool *self-optimizes* — a policy that picks the floored mechanism keeps re-picking it.

**Round 2** fixed two real problems with that round's evidence:

| Finding | Fix |
|---|---|
| **(A)** the proof used a non-production `--whale-seconds 600` override | **Conclusion changed.** The exact production command splits **nothing** at K=6 — neither whale crosses `total/K` (907.7s vs 814.0s / 660.2s). At K=6 the plan is `W/K`-bound at **15m8s with the whales whole**, still ahead of the ~19 min baseline. Count-sharding wins at **K≥8**, where the un-split makespan flattens at exactly llmhttp's 814.0s. |
| **(B)** the count-shard cost model was wrong-low | `countShardFloor` adds the observable fixed term and is now named, documented and reported as a **lower bound**. Each shard is a separate binary, so per-binary fixed work repeats S times — debaml's `sync.Once` witness is the case, and it is *invisible* to any `-json` reduction because it is billed to the test that triggers it. The bias (understating the shard cost ⇒ preferring count) is deliberate and documented. |
| **(C)** no regression at the workflow-equivalent total | `TestWhaleThresholdAtTheWorkflowScale` at ~5446s, with **default options** — a test needing an override is not a test of the production job. |
| **(D)** stale Phase-C bullet | Now describes the dominant-name comparison. |

**Round 3** closed the last two: **P1-1** made K one knob end to end (`TESTBUCKET_K` drives both `ingest --whale-k` and `plan --k`), with an end-to-end transition regression; **P3-1** aligned `defaultIngestOptions()` with the production CLI, including the `MinShardSeconds: 30` that was silently missing — tests claiming to exercise production were exercising a no-minimum variant.

### K is now genuinely one knob (round 3)

The split policy is derived and **stored by `ingest`**; `plan` only expands it. So `plan --k 8` on the K=6 store leaves both whales **whole** (makespan 814.0s = llmhttp's own weight), and a row recorded `count ×6` stays `×6` at K=8. The earlier "move to K=8" recommendation was therefore not realizable by a plan-side flag, and the record job leaned on the CLI `--whale-k` default — one knob in the brief, two in the YAML.

`TESTBUCKET_K` is now a single workflow-level value driving **both** `ingest --whale-k` and `plan --k`. The transition is one edit with a one-master-run settling time:

```
edit TESTBUCKET_K: "6" -> "8"
  -> next master record-timings re-records at the new K (llmhttp -> count x8)
  -> plan --k 8 realizes it
```

`TestKIsASingleKnobEndToEnd` walks all four steps and asserts the intermediate state too — that `plan --k 8` on a K=6 store emits *no* shards — so the model can't drift from the docs silently.

### The K sweep, correctly labelled

| K | whales whole (real) | count-sharded (**forced what-if**) | ideal `W/K` |
|---:|---:|---:|---:|
| 4 | 22m42s | 22m42s | 22m42s |
| 6 | **15m8s** ← production | 15m8s | 15m8s |
| 8 | 13m34s | 11m22s | 11m21s |
| 10 | 13m34s | 11m18s | 9m5s |
| 12 | 13m34s | 11m18s | 7m34s |

The middle column is a projection of what a re-record *would* produce. **The one K=8 number from the real transition:** `--whale-k 8` splits llmhttp ×8 (`-count=13`, aggregate 104 ≥ 100), leaves debaml whole at 660.2s under the new 680.8s threshold, and `plan --k 8` gives **681.8s = 11m22s at 0.1% imbalance**.

From K=8 the un-split column **flattens at exactly 814.0s — llmhttp's whole weight.** That is the floor count-sharding removes.

### K recommendation

llmhttp's weight agrees across this measurement and the workflow's comments (~13.6–15 min); **`W` does not** (90.8 min projected vs ~34 min implied), and `W` decides whether the whale binds at K=6.

**Ship Phase C at `TESTBUCKET_K=6` with the count-shard policy live**, and move the single knob to 8 only if the first master seed confirms the larger `W`. K=6 is safe under both branches; K=8 now bets on the least trustworthy number here. Because the knob is genuinely single, that move costs one edit plus one master run.

The qualitative divisibility answer is unchanged and remains **GO**.

## Scope

`.github/workflows/unit-tests.yml`, `scripts/testbucket-capture*.sh`, `scripts/testbucket-project-store.py`, and the `cmd/testbucket` split policy (`ingest.go`, `store.go`, `main.go`, `whales.go`). The `cmd/testbucket` change is what makes the divisibility finding actionable — without it the tool would emit the floored mechanism. Never-drop-a-test gate and grammar guard stay green (15/15). **Do not merge** — stop at the graph's gate. The store seeds on the first master run after merge; Phase C consumes it.

### CODE REVIEW

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -105,15 +105,41 @@ jobs:
...
             fi
           done
 
+      - name: Upload timing events
+        # if: always() so a RED run still contributes what it measured. ingest
+        # drops a failed package's weight (its wall time measures the failure,
+        # not the work) but keeps every package that passed, so one flaky lane
+        # cannot stall the store.
+        if: always()
+        uses: actions/upload-artifact@v7.0.1
```
[the remark this answers is not in the document]
Comment (invakid404) [reply .github/workflows/unit-tests.yml]: @coderabbitai Declined, with reasoning — and I'd be glad to do it as a separate change if the owner wants it.

The recommendation is sound in general: tags are mutable and SHAs are not. But this is a **repo-wide policy decision rather than a Phase B defect**, and applying it here would make things worse rather than better.

This repository pins every action to an exact version tag and has **zero SHA pins across 100+ references in eleven workflows**, with renovate keeping those tags current (`.github/workflows/renovate-sync.yml`). The steps this PR adds — `checkout@v7.0.0`, `setup-go@v6.5.0`, `upload-artifact@v7.0.1`, `download-artifact@v8.0.1`, `cache/restore@v4.2.4`, `cache/save@v4.2.4` — are the same actions at the same tags the repo already standardises on, and several of the lines you list (136, 346, 448, 458, 482, 487, 502) belong to the pre-existing `unit-tests`, `bamlutils-race` and `standalone-modules` jobs that this PR does not touch.

Pinning just this file would fragment the convention *and* take those references out of renovate's reach, so they would silently stop receiving security updates — trading one supply-chain risk for another. Doing it properly means all eleven workflows plus the renovate config to keep the digests moving, which is its own reviewable change with its own blast radius.

I've noted it in the PR as a follow-up candidate rather than silently dropping it.

File: .github/workflows/unit-tests.yml [hunk 2]
```diff
@@ -301,8 +327,222 @@ jobs:
...
+  #     weeks later as a mysteriously bad split.
+  record-timings:
+    needs: [unit-tests, bamlutils-race]
+    # !cancelled() rather than always(): record what a red run measured (the
+    # passing packages still have honest timings, and ingest drops the failed
+    # ones), but do not try to record a cancelled one, whose artifacts are
+    # arbitrary partial captures.
+    if: ${{ !cancelled() && github.event_name == 'push' && github.ref == 'refs/heads/master' }}
+    runs-on: ubuntu-latest
+    timeout-minutes: 15
```
[the remark this answers is not in the document]
Comment (invakid404) [reply .github/workflows/unit-tests.yml]: @coderabbitai Fixed in f89918d.

Added `permissions: { contents: read }` to both `record-timings` and `timings-restore`. Neither needs more: they read the tree, read/write the Actions cache, and handle artifacts. The cache and artifact APIs use the runtime token rather than `GITHUB_TOKEN` contents scope, so read is sufficient.

Worth noting this lines up with the design rather than just satisfying the linter: owner decision 3 deliberately keeps the timing store **out of the repo** (a rolling cache, not a bot commit) precisely so no CI job ever needs write access to contents. The permissions block makes that explicit at the job level. `permissions:` blocks already have precedent in this repo (`bamlfuzz-nightly.yml`, `integration-tests.yml`).

File: cmd/testbucket/ingest.go [hunk 3]
```diff
@@ -165,6 +165,82 @@ type ingestOptions struct {
...
+func chooseSplitPolicy(pkgSeconds, namedSeconds, heaviestName float64, namedCount, shards int) (policy, reason string) {
+	floor := countShardFloor(pkgSeconds, namedSeconds, shards)
+	switch {
+	case pkgSeconds <= 0 || shards < 2:
+		return splitCount, "no usable per-test picture"
+	case namedCount < 2:
+		return splitCount, "fewer than two named runnables to slice"
+	case namedSeconds/pkgSeconds < runUpgradeFraction:
+		return splitCount, fmt.Sprintf("named runnables explain only %.0f%% of the package (need %.0f%%)",
+			namedSeconds/pkgSeconds*100, runUpgradeFraction*100)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply cmd/testbucket/ingest.go]: @coderabbitai Fixed in f89918d.

Renamed to `heaviestSeconds`. You're right that it read as if a name were being passed — the caller keeps `heaviestName` for the test name and `heaviest` for the duration, so having the parameter called `heaviestName` while holding a `float64` was actively misleading at the one call site where the distinction matters.

File: cmd/testbucket/ingest.go [hunk 4]
```diff
@@ -382,16 +471,22 @@ func applyIngest(st *Store, sum *eventSummary, opt ingestOptions) (*ingestReport
...
-		if len(row.Tests) >= 2 && row.Seconds > 0 && named/row.Seconds >= runUpgradeFraction {
-			row.Split = splitRun
-		} else {
-			row.Split = splitCount
+		row.Split, row.SplitReason = chooseSplitPolicy(row.Seconds, named, heaviest, len(row.Tests), shards)
+		if row.Split == splitCount && heaviestName != "" {
+			rep.Dominated = append(rep.Dominated, fmt.Sprintf(
+				"%s: %s alone is %.0f%% of the package (%.1fs of %.1fs) — a -run split cannot finish faster than it",
+				pkg, heaviestName, heaviest/row.Seconds*100, heaviest, row.Seconds))
 		}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply cmd/testbucket/ingest.go]: @coderabbitai Fixed in f89918d — and this was a real correctness bug in the reporting, not a cosmetic one. Good catch.

You're right on all counts: `chooseSplitPolicy` returns `splitCount` for four different reasons and only one of them is dominance. A package sharded for "named runnables explain only 3% of the package" was being listed under **"count-sharded because one runnable dominates"** with the claim that a `-run` split cannot finish faster than that runnable — which for a 30s name in a 900s package is simply false, and would send a reviewer of the matrix looking for a problem that isn't there. And yes, the existing "per-test data explaining only a sliver" case produced exactly that state.

I took your cleaner alternative rather than the prefix match: `chooseSplitPolicy` now returns a discrete **cause** (`causeDominated`, `causeLowCoverage`, `causeTooFewNames`, `causeNoPicture`, `causeNameDivisible`) alongside the prose reason, and the caller gates on the cause. Matching on formatted text would have re-broken the moment someone reworded a reason string — which I have already done twice in this PR's review rounds.

`TestOnlyRealDominanceIsReportedAsDominance` builds all three count-sharded shapes (genuinely dominated, low named coverage, single named runnable), asserts all three are count-sharded, and asserts exactly one appears in the dominance report. The policy table also now pins the cause per case and cross-checks that `causeNameDivisible` and `splitRun` never disagree.

File: cmd/testbucket/store.go [hunk 5]
```diff
@@ -71,6 +71,11 @@ type UnitStat struct {
 	// test name, needs Tests). Set automatically by `ingest`.
 	Split     string `json:"split,omitempty"`
 	SplitInto int    `json:"split_into,omitempty"`
+	// SplitReason records WHY this policy was chosen. It is provenance, not
+	// input — nothing reads it back — but a store that says "count-sharded
+	// because TestX alone is 50% of the package" answers the question a
+	// reviewer of the emitted matrix will actually have.
+	SplitReason string `json:"split_reason,omitempty"`
```
[the remark this answers is not in the document]
Comment (invakid404) [reply cmd/testbucket/store.go]: @coderabbitai Fixed in f89918d.

Correct — I added `SplitReason` in a previous round and updated the assignment path but not the two un-flag paths, so a row that dropped below the threshold (or whose affordable width collapsed under 2) kept a justification for a split it was no longer doing. `writeWhaleReport` then printed that stale reason next to `policy none x0`: a store explaining a decision that no longer applies.

Both branches now clear it, with a comment saying why the reason travels with the policy.

`TestUnflaggingClearsTheSplitReason` covers both paths — a package that shrinks below the threshold, and one above the threshold but too small for slicing to pay for itself — plus a check that the whale report prints no dominance claim for an unsplit package.

File: cmd/testbucket/whales.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	store := fs.String("store", "test-timings.json", "timing store to analyse, or - for stdin")
+	k := fs.Int("k", 6, "bucket count the split threshold (total/K) is derived from")
+	top := fs.Int("top", 8, "how many runnables to list per package")
+	all := fs.Bool("all", false, "report every package with per-test rows, not just the split candidates")
+	if err := fs.Parse(args); err != nil {
+		return err
+	}
+	if *k < 1 {
+		return fmt.Errorf("--k must be >= 1, got %d", *k)
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply cmd/testbucket/whales.go]: @coderabbitai Fixed in f89918d, in both places you suggested.

`--top` is validated at the flag boundary (matching the `--k` check right above it), **and** `writeWhaleReport` clamps the limit from below as well as above. Both, because the report function is reachable directly from tests, so the flag check alone would leave the panic live on that path.

`TestWhaleReportSurvivesANegativeTop` drives `-1`, `-100`, `0` and `1` through the report with a `recover()` that fails the test on a panic, and asserts each still produces output.

File: scripts/testbucket-project-store.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    """Two-point fit, clamped: elapsed(c) = fixed + c*per_iter."""
+    per = max((high - low) / float(hi_c - lo_c), 0.0)
+    fixed = max(low - per * lo_c, 0.0)
+    return fixed + to_c * per
+
+
+def main():
+    ap = argparse.ArgumentParser()
+    ap.add_argument("--low", required=True); ap.add_argument("--low-count", type=int, required=True)
+    ap.add_argument("--high", required=True); ap.add_argument("--high-count", type=int, required=True)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply scripts/testbucket-project-store.py]: @coderabbitai Resolved by removing the file in f89918d.

`scripts/testbucket-project-store.py` was a one-off local measurement aid, not part of the Phase B deliverable. It fitted two laptop sweeps to a projected `-count=100` store so the divisibility question could be answered before any CI data existed; that answer now lives in the Phase B report, and the script is obsolete the moment the first master run lands real Ubuntu `-race -count=100` data — the measurement that actually drives the split.

Nothing in the repo referenced it, and the durable capability it was meant to provide is already committed and tested: `testbucket whales` prints the per-runnable distribution and both mechanism costs behind each split decision, and reads **any** store — including the `testbucket-timing-store` artifact a master `record-timings` run uploads. So the question stays re-answerable on CI's own numbers without maintaining a projection tool whose output the real data supersedes.

Your finding here was valid against the code as written; it is moot now that the code is gone rather than declined.

File: scripts/testbucket-project-store.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    ap.add_argument("--to-count", type=int, default=100)
+    ap.add_argument("--race", action="store_true", default=True)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply scripts/testbucket-project-store.py]: @coderabbitai Resolved by removing the file in f89918d.

`scripts/testbucket-project-store.py` was a one-off local measurement aid, not part of the Phase B deliverable. It fitted two laptop sweeps to a projected `-count=100` store so the divisibility question could be answered before any CI data existed; that answer now lives in the Phase B report, and the script is obsolete the moment the first master run lands real Ubuntu `-race -count=100` data — the measurement that actually drives the split.

Nothing in the repo referenced it, and the durable capability it was meant to provide is already committed and tested: `testbucket whales` prints the per-runnable distribution and both mechanism costs behind each split decision, and reads **any** store — including the `testbucket-timing-store` artifact a master `record-timings` run uploads. So the question stays re-answerable on CI's own numbers without maintaining a projection tool whose output the real data supersedes.

Your finding here was valid against the code as written; it is moot now that the code is gone rather than declined.

File: scripts/testbucket-project-store.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    for pkg in sorted(set(lo) & set(hi)):
+        lo_s, lo_t = lo[pkg]
+        hi_s, hi_t = hi[pkg]
+        seconds = fit(lo_s, hi_s, args.low_count, args.high_count, args.to_count)
+        if seconds <= 0:
+            continue
+        row = {"seconds": round(seconds, 3), "samples": 1}
+        projected = {}
+        for name in set(lo_t) | set(hi_t):
+            w = fit(lo_t.get(name, 0.0), hi_t.get(name, 0.0), args.low_count, args.high_count, args.to_count)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply scripts/testbucket-project-store.py]: @coderabbitai Resolved by removing the file in f89918d.

`scripts/testbucket-project-store.py` was a one-off local measurement aid, not part of the Phase B deliverable. It fitted two laptop sweeps to a projected `-count=100` store so the divisibility question could be answered before any CI data existed; that answer now lives in the Phase B report, and the script is obsolete the moment the first master run lands real Ubuntu `-race -count=100` data — the measurement that actually drives the split.

Nothing in the repo referenced it, and the durable capability it was meant to provide is already committed and tested: `testbucket whales` prints the per-runnable distribution and both mechanism costs behind each split decision, and reads **any** store — including the `testbucket-timing-store` artifact a master `record-timings` run uploads. So the question stays re-answerable on CI's own numbers without maintaining a projection tool whose output the real data supersedes.

Your finding here was valid against the code as written; it is moot now that the code is gone rather than declined.

File: scripts/testbucket-project-store.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            for pkg in sorted(units):
+                row = units[pkg]
+                for name in sorted(row.get("tests", {})):
+                    fh.write(json.dumps({"Action": "pass", "Package": pkg,
+                                         "Test": name, "Elapsed": row["tests"][name]}) + "\n")
+                fh.write(json.dumps({"Action": "pass", "Package": pkg,
+                                     "Elapsed": row["seconds"]}) + "\n")
+    total = sum(u["seconds"] for u in units.values())
+    print("projected %d packages to %s: total %.1fs (%.1f min) -> %s"
+          % (len(units), flags, total, total / 60.0, args.out), file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply scripts/testbucket-project-store.py]: @coderabbitai Resolved by removing the file in f89918d.

`scripts/testbucket-project-store.py` was a one-off local measurement aid, not part of the Phase B deliverable. It fitted two laptop sweeps to a projected `-count=100` store so the divisibility question could be answered before any CI data existed; that answer now lives in the Phase B report, and the script is obsolete the moment the first master run lands real Ubuntu `-race -count=100` data — the measurement that actually drives the split.

Nothing in the repo referenced it, and the durable capability it was meant to provide is already committed and tested: `testbucket whales` prints the per-runnable distribution and both mechanism costs behind each split decision, and reads **any** store — including the `testbucket-timing-store` artifact a master `record-timings` run uploads. So the question stays re-answerable on CI's own numbers without maintaining a projection tool whose output the real data supersedes.

Your finding here was valid against the code as written; it is moot now that the code is gone rather than declined.
